### PR TITLE
ci: automated release pipeline

### PR DESCRIPTION
**Pull request by malt3**
### Proposed change(s)

- [AB#2722](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/2722) ci: automated release pipeline

### Additional info

- Full pipeline run: https://github.com/edgelesssys/constellation/actions/runs/3900199072

### Checklist

- [x] Update release checklist to reference pipeline
- [x] Add labels (e.g., for changelog category)

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+          ref: ${{ github.head_ref }}
+      - name: Create release branch
+        run: |
+          git fetch
+          git pull
+          git checkout "${BRANCH}" || git checkout -B "${BRANCH}"
+          git push origin "${BRANCH}"
+
+  micro-services:
+    needs: [verify-inputs, prepare-release-branch]
```
Comment (katexochen) [inline .github/workflows/release.yml:118]: Does this work if prepare-release-branch is skipped?

File: internal/attestation/measurements/measurement-generator/generate.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+	version, err := versionsapi.NewVersionFromShortPath(image, versionsapi.VersionKindImage)
+	if err != nil {
+		return nil, fmt.Errorf("parsing image name: %w", err)
+	}
+	url, err := url.Parse(constants.CDNRepositoryURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing image version repository URL: %w", err)
+	}
+	url.Path = path.Join(constants.CDNAPIPrefix, "ref", version.Ref, "stream", version.Stream, version.Version, "image", "csp", strings.ToLower(provider.String()), file)
+	return url, nil
```
Comment (katexochen) [inline internal/attestation/measurements/measurement-generator/generate.go:78]: use `version.ArtifactURL()`

File: internal/attestation/measurements/measurement-generator/generate.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+		Type: &ast.Ident{
+			NamePos: pos,
+			Name:    "M",
+		},
+		Elts:   elts,
+		Rbrace: entryPos - 1,
+	}
+}
+
+func main() {
```
Comment (katexochen) [inline internal/attestation/measurements/measurement-generator/generate.go:199]: func main should be at the top of the file

File: internal/attestation/measurements/measurement-generator/generate.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+				return true
+			}
+			returnStmt, ok := clause.Body[0].(*ast.ReturnStmt)
+			if !ok || len(returnStmt.Results) == 0 {
+				return true
+			}
+
+			provider := cloudprovider.FromString(sel.Sel.Name)
+			if provider == cloudprovider.Unknown {
+				return true
```
Comment (katexochen) [inline internal/attestation/measurements/measurement-generator/generate.go:230]: log that an unknown provider was found

File: internal/attestation/measurements/measurement-generator/generate.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+	return &ast.KeyValueExpr{
+		Key: &ast.BasicLit{
+			ValuePos: pos,
+			Kind:     token.INT,
+			Value:    key,
+		},
+		Colon: colon,
+		Value: &ast.CompositeLit{
+			// Type:   ast.NewIdent("Measurement"),
```
Comment (katexochen) [inline internal/attestation/measurements/measurement-generator/generate.go]: Why is this line commented?

File: internal/attestation/measurements/measurement-generator/generate.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	return &ast.CompositeLit{
+		Lbrace: lbrace,
+		Type: &ast.Ident{
+			NamePos: pos,
+			Name:    "M",
+		},
+		Elts:   elts,
+		Rbrace: entryPos - 1,
+	}
+}
```
Comment (katexochen) [inline internal/attestation/measurements/measurement-generator/generate.go:197]: Please add a decent amount of documentation to this code, so someone who changed a thing in the target file can update this code in a reasonable time.

Comment (katexochen) [reply internal/attestation/measurements/measurement-generator/generate.go:197]: Also cool stuff by the way...

File: internal/attestation/measurements/measurement-generator/generate.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+				return true
+			}
+
+			provider := cloudprovider.FromString(sel.Sel.Name)
+			if provider == cloudprovider.Unknown {
+				return true
+			}
+			log.Println("Found", provider)
+			measuremnts := mustGetMeasurements(ctx, rekor, []byte(constants.CosignPublicKey), http.DefaultClient, provider, defaultConf.Image)
+			returnStmt.Results[0] = measurementsCompositeLiteral(measuremnts, returnStmt.Return+7)
```
Comment (katexochen) [inline internal/attestation/measurements/measurement-generator/generate.go:234]: Please add some type of feedback like a counter so we know if we found any measurements at all. Have a look at the updated hash-generator.

File: internal/attestation/measurements/measurements_enterprise.go [hunk 8]
```diff
@@ -10,6 +10,9 @@ package measurements
 
 import "github.com/edgelesssys/constellation/v2/internal/cloud/cloudprovider"
 
+// Regenerate the measurements by running go generate
```
Comment (katexochen) [inline internal/attestation/measurements/measurements_enterprise.go:13]: Add a comment why we need enterprise build tag

**Review by katexochen — CHANGES_REQUESTED** [review]
Reviewed commits including 0e82b96a133969f124884227cbad2d6d706191ad

File: .github/workflows/release.yml (hunk 1 above)
Comment (malt3) [reply .github/workflows/release.yml:118]: https://github.com/edgelesssys/constellation/actions/runs/3895082895 :disappointed:

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+          {
+            echo "WITHOUT_V=${WITHOUT_V}"
+            echo "PART_MAJOR=${PART_MAJOR}"
+            echo "PART_MINOR=${PART_MINOR}"
+            echo "PART_PATCH=${PART_PATCH}"
+            echo "MAJOR=${PART_MAJOR}"
+            echo "MAJOR_MINOR=${PART_MAJOR}.${PART_MINOR}"
+            echo "MAJOR_MINOR_PATCH=${PART_MAJOR}.${PART_MINOR}.${PART_PATCH}"
+            echo "RELEASE_BRANCH=release/v${PART_MAJOR}.${PART_MINOR}"
+          } >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (malt3) [reply .github/workflows/release.yml]: Fixed using tee

File: dev-docs/workflows/release.md [hunk 10]
```diff
@@ -1,140 +1,49 @@
 # Release Checklist
 
-This checklist will prepare `v1.3.0` from `v1.2.0`. Adjust your version numbers accordingly.
+This checklist will prepare `v1.3.0`  from `v1.2.0` (minor release) or `v1.3.1` from `v1.3.0` (patch release). Adjust your version numbers accordingly.
```
Comment (katexochen) [inline dev-docs/workflows/release.md:3]:
```suggestion
This checklist will prepare `v1.3.0` from `v1.2.0` (minor release) or `v1.3.1` from `v1.3.0` (patch release). Adjust your version numbers accordingly.
```
nit

**Review by katexochen — COMMENTED** [review]
Reviewed including abe99fb47d89320d9b8ec0dd8524cdec885003c3

**Review by daniel-weisse — APPROVED** [review]
Havent really reviewed the CI part, but measurements updates look good to me

File: internal/attestation/measurements/measurement-generator/generate.go [hunk 11]
```diff
@@ -32,6 +32,9 @@ import (
 	"golang.org/x/tools/go/ast/astutil"
 )
 
+// this tool is used to generate hardcoded measurements for the enterprise build.
+// Measurements are embedded in the constellation cli.
```
Comment (katexochen) [inline internal/attestation/measurements/measurement-generator/generate.go:36]:
```suggestion
// This tool is used to generate hard-coded measurements for the enterprise build.
// Measurements are embedded in the Constellation CLI.
```
This should be a package comment.

**Review by katexochen — COMMENTED** [review]
Reviewed including 7cca6dc35f4bc9211b8f1b3565e2dc0f1e1b3e85

**Review by katexochen — APPROVED** [review]
:rocket:

**Pull request merged**